src/cnxml/whitespace.js

```diff
@@ -0,0 +1,279 @@
+
+    let before, after
+
+    const first = node[0]
+
+    if (first && first.object === 'text') {
+        const m = first.text.match(/^\s+/u)
+
+        if (m !== null) {
+            before = m[0]
+            first.text = first.text.slice(before.length)
+        }
+    } else if (first) {
+        const [b, a] = adjustWS(first)
+        before = b
+
+        if (a) {
+            if (node[1] && node[1].object === 'text') {
+                node[1].text = a + node[1].text
+            } else {
+                node.splice(1, 0, {
+                    object: 'text',
+                    text: a,
+                })
+            }
+        }
+    }
+
+    const last = node[node.length - 1]
+    const lastlast = node[node.length - 2]
+
+    if (last && last.object === 'text') {
+        const m = last.text.match(/\s+$/u)
+
+        if (m !== null) {
+            after = m[0]
+            last.text = last.text.slice(0, m.index)
+        }
+    } else if (last) {
+        const [b, a] = adjustWS(last)
+        after = a
+
+        if (b) {
+            if (lastlast && lastlast.object === 'text') {
+                lastlast.text += b
+            } else {
+                node.splice(node.length - 1, 0, {
+                    object: 'text',
+                    text: b,
+                })
+            }
+        }
+    }
+
+    for (let i=0 ; i<node.length ; ++i) {
+        if (node[i].object === 'text' && node[i].text.length === 0) {
+            node.splice(i, 1)
+            i -= 1
+        }
+    }
+
+    return [before, after]
+}
+
+/**
+ * Simplify white space by replacing certain groups of code points with a single
+ * codepoint.
+ *
+ * This function performs steps 1 through 6, assuming that step 7 has already
+ * been performed.
+ */
+function replaceWS(node) {
+    if (node instanceof Array) {
+        for (const child of node) {
+            replaceWS(child)
+        }
+        return
+    }
+
+    if (typeof node === 'string') {
+        // 1st step
+        return node.replace(/\s/gu, replaceWSChar)
+            // 2nd step
+            .replace(/\s[\u180e\u200b\u200c\u200d\u2060]/g, c => c[0])
+            .replace(/[\u180e\u200b\u200c\u200d\u2060]\s/g, c => c[1])
+            // 3rd step
+            .replace(/[\s\u180e\u200b\u200c\u200d\u2060]{2,}/g, collapseWSSequence)
+            // 4th step
+            .replace(/[\u0020\u2000-\u2006\u2008\u2009\u200A\u205F]+([\u1680\u3000])/g, (_, r) => r)
+            .replace(/([\u1680\u3000][\u0020\u2000-\u2006\u2008\u2009\u200A\u205F]+)/g, (_, r) => r)
+            // 5th step
+            .replace(/[\u0020\u1680\u2000-\u2006\u2008\u2009\u200A\u205F\u3000]+([\u00a0])/g, (_, r) => r)
+            .replace(/([\u00a0][\u0020\u1680\u2000-\u2006\u2008\u2009\u200A\u205F\u3000]+)/g, (_, r) => r)
+            // 6th step
+            .replace(/[\s\u180e\u200b\u200c\u200d\u2060]{2,}/g, c => c[0])
+    }
+
+    switch (node.object) {
+    case 'text':
+        node.text = replaceWS(node.text)
```

Comment:
This is failing for:
```
Object { object: "text", text: "<m:math xmlns:m=\"http://www.w3.org/1998/Math/MathML\" overflow=\"scroll\"><m:mrow><m:mi>A</m:mi><m:mo>,</m:mo><m:mspace width=\"0.166667em\"/><m:mi>B</m:mi><m:mo>,</m:mo><m:mspace width=\"0.166667em\"/><m:mi>C</m:mi></m:mrow></m:math>", marks: [] }
```
`Error: Cannot set on an immutable record.`

Reply:
This is because `node` is created as immutable object instead of normal js object: https://github.com/katalysteducation/adaptarr-front/blob/master/src/screens/app/Draft/plugins/SourceElements/deserialize.ts#L15